File: frontend/src/services/studentPortfolio.ts
import api from "../lib/api";

export type ActivityCategory = "club" | "sport" | "society" | "competition";
export type DisciplinarySeverity = "minor" | "major" | "severe";

export const ACTIVITY_CATEGORIES: { value: ActivityCategory; label: string }[] = [
  { value: "club", label: "Club" },
  { value: "sport", label: "Sport" },
  { value: "society", label: "Society" },
  { value: "competition", label: "Competition" },
];

export const DISCIPLINARY_SEVERITIES: { value: DisciplinarySeverity; label: string }[] = [
  { value: "minor", label: "Minor" },
  { value: "major", label: "Major" },
  { value: "severe", label: "Severe" },
];

// Matches db.ListProgressReportsByStudentRow
export interface ProgressReport {
  id: string;
  student_id: string;
  term_id: string;
  narrative: string;
  written_by: string | null;
  created_at: string | null;
  updated_at: string | null;
  term_name: string;
}

export interface CreateProgressReportRequest {
  term_id: string;
  narrative: string;
}

// Matches db.StudentActivity
export interface StudentActivity {
  id: string;
  student_id: string;
  academic_year_id: string;
  category: ActivityCategory;
  name: string;
  role: string | null;
  achievement: string | null;
  created_at: string | null;
}

export interface CreateActivityRequest {
  academic_year_id: string;
  category: ActivityCategory;
  name: string;
  role?: string;
  achievement?: string;
}

export type UpdateActivityRequest = Omit<CreateActivityRequest, "academic_year_id">;

// Matches db.StudentLeadershipRole
export interface LeadershipRole {
  id: string;
  student_id: string;
  academic_year_id: string;
  title: string;
  scope: string | null;
  created_at: string | null;
}

export interface CreateLeadershipRoleRequest {
  academic_year_id: string;
  title: string;
  scope?: string;
}

// Matches db.StudentAward
export interface StudentAward {
  id: string;
  student_id: string;
  academic_year_id: string;
  title: string;
  category: string | null;
  awarded_date: string;
  description: string | null;
  created_at: string | null;
}

export interface CreateAwardRequest {
  academic_year_id: string;
  title: string;
  category?: string;
  awarded_date: string; // RFC3339 timestamp
  description?: string;
}

// Matches db.StudentDisciplinaryRecord
export interface DisciplinaryRecord {
  id: string;
  student_id: string;
  academic_year_id: string;
  incident_date: string;
  description: string;
  action_taken: string | null;
  severity: DisciplinarySeverity;
  recorded_by: string | null;
  created_at: string | null;
}

export interface CreateDisciplinaryRecordRequest {
  academic_year_id: string;
  incident_date: string; // RFC3339 timestamp
  description: string;
  action_taken?: string;
  severity: DisciplinarySeverity;
}

// Matches db.ListPrefectAppointmentsByStudentRow
export interface PrefectAppointment {
  id: string;
  academic_year_id: string;
  rank: string;
  created_at: string | null;
  academic_year_label: string;
}

export const studentPortfolioApi = {
  listProgressReports: (studentId: string) =>
    api.get<ProgressReport[]>(`/students/${studentId}/progress-reports`).then((r) => r.data),
  createProgressReport: (studentId: string, data: CreateProgressReportRequest) =>
    api.post<ProgressReport>(`/students/${studentId}/progress-reports`, data).then((r) => r.data),
  updateProgressReport: (studentId: string, recordId: string, narrative: string) =>
    api
      .put<ProgressReport>(`/students/${studentId}/progress-reports/${recordId}`, { narrative })
      .then((r) => r.data),
  deleteProgressReport: (studentId: string, recordId: string) =>
    api.delete(`/students/${studentId}/progress-reports/${recordId}`).then((r) => r.data),

  listActivities: (studentId: string) =>
    api.get<StudentActivity[]>(`/students/${studentId}/activities`).then((r) => r.data),
  createActivity: (studentId: string, data: CreateActivityRequest) =>
    api.post<StudentActivity>(`/students/${studentId}/activities`, data).then((r) => r.data),
  deleteActivity: (studentId: string, recordId: string) =>
    api.delete(`/students/${studentId}/activities/${recordId}`).then((r) => r.data),

  listLeadershipRoles: (studentId: string) =>
    api.get<LeadershipRole[]>(`/students/${studentId}/leadership-roles`).then((r) => r.data),
  createLeadershipRole: (studentId: string, data: CreateLeadershipRoleRequest) =>
    api.post<LeadershipRole>(`/students/${studentId}/leadership-roles`, data).then((r) => r.data),
  deleteLeadershipRole: (studentId: string, recordId: string) =>
    api.delete(`/students/${studentId}/leadership-roles/${recordId}`).then((r) => r.data),

  listAwards: (studentId: string) =>
    api.get<StudentAward[]>(`/students/${studentId}/awards`).then((r) => r.data),
  createAward: (studentId: string, data: CreateAwardRequest) =>
    api.post<StudentAward>(`/students/${studentId}/awards`, data).then((r) => r.data),
  deleteAward: (studentId: string, recordId: string) =>
    api.delete(`/students/${studentId}/awards/${recordId}`).then((r) => r.data),

  listDisciplinaryRecords: (studentId: string) =>
    api.get<DisciplinaryRecord[]>(`/students/${studentId}/disciplinary-records`).then((r) => r.data),
  createDisciplinaryRecord: (studentId: string, data: CreateDisciplinaryRecordRequest) =>
    api.post<DisciplinaryRecord>(`/students/${studentId}/disciplinary-records`, data).then((r) => r.data),
  deleteDisciplinaryRecord: (studentId: string, recordId: string) =>
    api.delete(`/students/${studentId}/disciplinary-records/${recordId}`).then((r) => r.data),

  listPrefectAppointments: (studentId: string) =>
    api
      .get<PrefectAppointment[]>(`/students/${studentId}/prefect-appointments`)
      .then((r) => r.data),
};

File: backend/internal/services/student_portfolio.go
package services

import (
	"context"
	"errors"
	"fmt"

	"github.com/google/uuid"
	"github.com/jackc/pgx/v5"
	db "github.com/openschool-org/openschool/db/sqlc"
	"github.com/openschool-org/openschool/internal/models"
	"github.com/openschool-org/openschool/internal/repositories"
)

var (
	ErrInvalidActivityCategory     = errors.New("invalid activity category")
	ErrInvalidDisciplinarySeverity = errors.New("invalid severity")
	ErrPortfolioRecordNotFound     = errors.New("record not found")
)

type StudentPortfolioService struct {
	repo        *repositories.StudentPortfolioRepository
	teacherRepo *repositories.TeacherRepository
}

func NewStudentPortfolioService(repo *repositories.StudentPortfolioRepository, teacherRepo *repositories.TeacherRepository) *StudentPortfolioService {
	return &StudentPortfolioService{repo: repo, teacherRepo: teacherRepo}
}

// TeacherProfileIDForUser resolves the signed-in caller's teacher_profiles
// id (nil if they're not a teacher, e.g. an admin) — used to credit
// "written_by" on a progress report without exposing the teacher lookup to
// the handler layer.
func (s *StudentPortfolioService) TeacherProfileIDForUser(ctx context.Context, userID uuid.UUID) *uuid.UUID {
	teacher, err := s.teacherRepo.GetByUserID(ctx, userID)
	if err != nil {
		return nil
	}
	return &teacher.ID
}

// Progress reports

func (s *StudentPortfolioService) CreateProgressReport(ctx context.Context, studentID uuid.UUID, req models.CreateProgressReportRequest, writtenBy *uuid.UUID) (db.StudentProgressReport, error) {
	termID, err := uuid.Parse(req.TermID)
	if err != nil {
		return db.StudentProgressReport{}, fmt.Errorf("invalid term id")
	}
	return s.repo.CreateProgressReport(ctx, studentID, termID, req.Narrative, writtenBy)
}

func (s *StudentPortfolioService) ListProgressReports(ctx context.Context, studentID uuid.UUID) ([]db.ListProgressReportsByStudentRow, error) {
	return s.repo.ListProgressReports(ctx, studentID)
}

func (s *StudentPortfolioService) UpdateProgressReport(ctx context.Context, id, studentID uuid.UUID, req models.UpdateProgressReportRequest) (db.StudentProgressReport, error) {
	report, err := s.repo.UpdateProgressReport(ctx, id, studentID, req.Narrative)
	if errors.Is(err, pgx.ErrNoRows) {
		return db.StudentProgressReport{}, ErrPortfolioRecordNotFound
	}
	return report, err
}

func (s *StudentPortfolioService) DeleteProgressReport(ctx context.Context, id, studentID uuid.UUID) error {
	rows, err := s.repo.DeleteProgressReport(ctx, id, studentID)
	if err != nil {
		return err
	}
	if rows == 0 {
		return ErrPortfolioRecordNotFound
	}
	return nil
}

// Activities

func (s *StudentPortfolioService) CreateActivity(ctx context.Context, studentID uuid.UUID, req models.CreateActivityRequest) (db.StudentActivity, error) {
	if !models.ValidActivityCategories[req.Category] {
		return db.StudentActivity{}, ErrInvalidActivityCategory
	}
	yearID, err := uuid.Parse(req.AcademicYearID)
	if err != nil {
		return db.StudentActivity{}, fmt.Errorf("invalid academic year id")
	}
	return s.repo.CreateActivity(ctx, studentID, yearID, req.Category, req.Name, req.Role, req.Achievement)
}

func (s *StudentPortfolioService) ListActivities(ctx context.Context, studentID uuid.UUID) ([]db.StudentActivity, error) {
	return s.repo.ListActivities(ctx, studentID)
}

func (s *StudentPortfolioService) UpdateActivity(ctx context.Context, id, studentID uuid.UUID, req models.UpdateActivityRequest) (db.StudentActivity, error) {
	if !models.ValidActivityCategories[req.Category] {
		return db.StudentActivity{}, ErrInvalidActivityCategory
	}
	activity, err := s.repo.UpdateActivity(ctx, id, studentID, req.Category, req.Name, req.Role, req.Achievement)
	if errors.Is(err, pgx.ErrNoRows) {
		return db.StudentActivity{}, ErrPortfolioRecordNotFound
	}
	return activity, err
}

func (s *StudentPortfolioService) DeleteActivity(ctx context.Context, id, studentID uuid.UUID) error {
	rows, err := s.repo.DeleteActivity(ctx, id, studentID)
	if err != nil {
		return err
	}
	if rows == 0 {
		return ErrPortfolioRecordNotFound
	}
	return nil
}

// Leadership roles

func (s *StudentPortfolioService) CreateLeadershipRole(ctx context.Context, studentID uuid.UUID, req models.CreateLeadershipRoleRequest) (db.StudentLeadershipRole, error) {
	yearID, err := uuid.Parse(req.AcademicYearID)
	if err != nil {
		return db.StudentLeadershipRole{}, fmt.Errorf("invalid academic year id")
	}
	return s.repo.CreateLeadershipRole(ctx, studentID, yearID, req.Title, req.Scope)
}

func (s *StudentPortfolioService) ListLeadershipRoles(ctx context.Context, studentID uuid.UUID) ([]db.StudentLeadershipRole, error) {
	return s.repo.ListLeadershipRoles(ctx, studentID)
}

func (s *StudentPortfolioService) DeleteLeadershipRole(ctx context.Context, id, studentID uuid.UUID) error {
	rows, err := s.repo.DeleteLeadershipRole(ctx, id, studentID)
	if err != nil {
		return err
	}
	if rows == 0 {
		return ErrPortfolioRecordNotFound
	}
	return nil
}

// Awards

func (s *StudentPortfolioService) CreateAward(ctx context.Context, studentID uuid.UUID, req models.CreateAwardRequest) (db.StudentAward, error) {
	yearID, err := uuid.Parse(req.AcademicYearID)
	if err != nil {
		return db.StudentAward{}, fmt.Errorf("invalid academic year id")
	}
	return s.repo.CreateAward(ctx, studentID, yearID, req.Title, req.Category, req.AwardedDate, req.Description)
}

func (s *StudentPortfolioService) ListAwards(ctx context.Context, studentID uuid.UUID) ([]db.StudentAward, error) {
	return s.repo.ListAwards(ctx, studentID)
}

func (s *StudentPortfolioService) DeleteAward(ctx context.Context, id, studentID uuid.UUID) error {
	rows, err := s.repo.DeleteAward(ctx, id, studentID)
	if err != nil {
		return err
	}
	if rows == 0 {
		return ErrPortfolioRecordNotFound
	}
	return nil
}

// Disciplinary records

func (s *StudentPortfolioService) CreateDisciplinaryRecord(ctx context.Context, studentID uuid.UUID, req models.CreateDisciplinaryRecordRequest, recordedBy *uuid.UUID) (db.StudentDisciplinaryRecord, error) {
	if !models.ValidDisciplinarySeverities[req.Severity] {
		return db.StudentDisciplinaryRecord{}, ErrInvalidDisciplinarySeverity
	}
	yearID, err := uuid.Parse(req.AcademicYearID)
	if err != nil {
		return db.StudentDisciplinaryRecord{}, fmt.Errorf("invalid academic year id")
	}
	return s.repo.CreateDisciplinaryRecord(ctx, studentID, yearID, req.IncidentDate, req.Description, req.ActionTaken, req.Severity, recordedBy)
}

func (s *StudentPortfolioService) ListDisciplinaryRecords(ctx context.Context, studentID uuid.UUID) ([]db.StudentDisciplinaryRecord, error) {
	return s.repo.ListDisciplinaryRecords(ctx, studentID)
}

func (s *StudentPortfolioService) DeleteDisciplinaryRecord(ctx context.Context, id, studentID uuid.UUID) error {
	rows, err := s.repo.DeleteDisciplinaryRecord(ctx, id, studentID)
	if err != nil {
		return err
	}
	if rows == 0 {
		return ErrPortfolioRecordNotFound
	}
	return nil
}
